File: ros/learning_pkg_python/learning_pkg_python/topic_sub.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class SubNode(Node):
    def __init__(self):
        super().__init__('topic_sub')
        # 订阅名为'chatter'的话题，消息类型为String
        self.sub = self.create_subscription(
            String,
            'chatter',
            self.callback,  # 收到消息时执行的函数
            10  # 队列大小
        )

    def callback(self, msg):
        # 打印收到的消息内容
        self.get_logger().info(f'Received: {msg.data}')

def main(args=None):
    rclpy.init(args=args)
    node = SubNode()
    rclpy.spin(node)  # 保持节点运行，等待消息
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: ros/learning_pkg_cpp/src/topic_sub.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class SubNode : public rclcpp::Node
{
public:
    SubNode() : Node("topic_sub_cpp")
    {
        subscriber_ = this->create_subscription<std_msgs::msg::String>(
            "chatter",
            10,
            std::bind(&SubNode::sub_callback, this, std::placeholders::_1)
        );
        
        RCLCPP_INFO(this->get_logger(), "C++ 订阅者节点已启动！等待消息...");
    }

private:
    void sub_callback(const std_msgs::msg::String::SharedPtr msg) const
    {
        RCLCPP_INFO(this->get_logger(), "收到: %s", msg->data.c_str());
    }

    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscriber_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SubNode>());
    rclcpp::shutdown();
    return 0;
}